cycle detected while resolving 'fromUcs4' in swift_name attribute for 'R' [#ClangDeclarationImport]

Trying out Swift C++ interopt with Qt, I’m getting a warning for this piece of code in Qt:

```c++
[[nodiscard]] constexpr auto QChar::fromUcs4(char32_t c) noexcept
{
    struct R {
        char16_t chars[2];
        [[nodiscard]] constexpr operator QStringView() const noexcept { return {begin(), end()}; }
        [[nodiscard]] constexpr qsizetype size() const noexcept { return chars[1] ? 2 : 1; }
        [[nodiscard]] constexpr const char16_t *begin() const noexcept { return chars; }
        [[nodiscard]] constexpr const char16_t *end() const noexcept { return begin() + size(); }
    };
    return requiresSurrogates(c) ? R{{QChar::highSurrogate(c),
                                      QChar::lowSurrogate(c)}} :
                                   R{{char16_t(c), u'\0'}} ;
}
```

```
/Users/torarne/build/qt/6.x-prefix/install/lib/QtCore.framework/Headers/qstringview.h:482:12: warning: cycle detected while resolving 'fromUcs4' in swift_name attribute for 'R' [#ClangDeclarationImport]
480 | [[nodiscard]] constexpr auto QChar::fromUcs4(char32_t c) noexcept
481 | {
482 |     struct R {
    |            |- warning: cycle detected while resolving 'fromUcs4' in swift_name attribute for 'R' [#ClangDeclarationImport]
    |            `- note: please report this issue to the owners of 'QtCore'
483 |         char16_t chars[2];
484 |         [[nodiscard]] constexpr operator QStringView() const noexcept { return {begin(), end()}; }
```

Based on @egorzhdan's comment in https://forums.swift.org/t/cycle-detected-while-resolving-swift-name/82025 this is a bug in the compiler. 

I'm still seeing it with `swift-driver version: 1.167 Apple Swift version 6.4 (swiftlang-6.4.0.20.104 clang-2100.3.20.102)`.

Is there a workaround we can do in the meantime?

Thanks!

--- a/lib/ClangImporter/ImportDecl.cpp
+++ b/lib/ClangImporter/ImportDecl.cpp
@@ -10560,6 +10560,11 @@ DeclContext *ClangImporter::Implementation::importDeclContextImpl(
   // TranslationUnit DeclContext here.
   assert(!dc->isTranslationUnit());
 
+  // A declaration declared inside a function body (e.g. a local C++ class) has
+  // no Swift decl context to be imported into.
+  if (dc->isFunctionOrMethod())
+    return nullptr;
+
   auto decl = dyn_cast<clang::NamedDecl>(dc);
   if (!decl || !decl->getDeclName().isIdentifier())
     return nullptr;

--- a/test/Interop/Cxx/class/Inputs/local-class.h
+++ b/test/Interop/Cxx/class/Inputs/local-class.h
@@ -1,0 +1,43 @@
+struct DeducedReturnType {
+  static constexpr auto returnsLocalClass(int c) noexcept {
+    struct Local {
+      char chars[2];
+    };
+    return Local{{(char)c, '\0'}};
+  }
+};
+
+inline auto localClassInFreeFunction() {
+  struct Local {
+    int x;
+  };
+  return Local{42};
+}
+
+inline auto nestedLocalClass() {
+  struct LocalOuter {
+    struct LocalInner {
+      int x;
+    };
+    LocalInner inner;
+  };
+  return LocalOuter{{42}};
+}
+
+inline auto localClassInLambda() {
+  auto fn = []() {
+    struct LocalInLambda {
+      int x;
+    };
+    return LocalInLambda{42};
+  };
+  return fn();
+}
+
+// Records nested at non-local scope should still be imported.
+struct Outer {
+  struct Inner {
+    int x;
+  };
+  Inner inner;
+};

--- a/test/Interop/Cxx/class/Inputs/module.modulemap
+++ b/test/Interop/Cxx/class/Inputs/module.modulemap
@@ -121,6 +121,11 @@ module InvalidNestedStruct {
   requires cplusplus
 }
 
+module LocalClass {
+  header "local-class.h"
+  requires cplusplus
+}
+
 module ClassProtocolNameClash {
   header "class-protocol-name-clash.h"
   requires cplusplus

--- a/test/Interop/Cxx/class/local-class-module-interface.swift
+++ b/test/Interop/Cxx/class/local-class-module-interface.swift
@@ -1,0 +1,6 @@
+// RUN: %target-swift-ide-test -print-module -module-to-print=LocalClass -I %S/Inputs/ -source-filename=x -cxx-interoperability-mode=default 2>&1 | %FileCheck %s
+
+// Classes declared inside a function body are not imported.
+
+// CHECK-NOT: warning
+// CHECK-NOT: struct Local

